query/src/sql/planner/binder/select.rs

```diff
@@ -193,4 +217,133 @@ impl<'a> Binder {
         let new_expr = SExpr::create_unary(filter_plan.into(), child);
         Ok(new_expr)
     }
+
+    pub(super) async fn bind_set_operator(
+        &mut self,
+        bind_context: &BindContext,
+        left: &SetExpr<'_>,
+        right: &SetExpr<'_>,
+        op: &SetOperator,
+        all: &bool,
+    ) -> Result<(SExpr, BindContext)> {
+        let (left_expr, mut left_bind_context) =
+            self.bind_set_expr(bind_context, &*left, &[]).await?;
+        let (right_expr, mut right_bind_context) =
+            self.bind_set_expr(bind_context, &*right, &[]).await?;
+        if left_bind_context.columns.len() != right_bind_context.columns.len() {
+            return Err(ErrorCode::SemanticError(
+                "SetOperation must have the same number of columns",
+            ));
+        } else {
+            for (left_col, right_col) in left_bind_context
+                .columns
+                .iter_mut()
+                .zip(right_bind_context.columns.iter_mut())
+            {
+                let left_data_type = remove_nullable(&left_col.data_type);
+                let right_data_type = remove_nullable(&right_col.data_type);
+                let left_id = left_data_type.data_type_id();
+                let right_id = right_data_type.data_type_id();
+                if left_id.is_numeric() && right_id.is_numeric() {
+                    let coercion_type =
+                        numerical_coercion(&left_data_type, &right_data_type, false)?;
+                    left_col.data_type = coercion_type.clone();
+                    right_col.data_type = coercion_type;
```

Comment:
Maybe we should keep the nullability of the coercion type?

Reply:
I'm not sure about that, I think both are ok. cc @sundy-li

Reply:
I think it's fine to use `merge_types` here, maybe @sundy-li would answer this.

Reply:
Let's use `merge_types`.

Reply:
done